doc/SAI_failure_handling/SAI_failure_handling.md

```diff
@@ -0,0 +1,123 @@
+
+2. Return `task_failed` -- No crash, no retry, although the failure is not handled successfully, orchagent will keep running without exit.
+
+The failure handling function should return `task_failed` when the failure is unable to be handled in orchagent and another attempt is not likely to resolve the failure. In such a scenario, the function should prevent orchagent from retrying and escalate the failure to upper layers. Example scenarios where the functions should return this status include invalid user input (e.g., a wrong ACL, conflicting IP addresses), hardware permanent error, non-critical internal logic error, etc.
+
+3. Return `task_need_retry` --  No crash, retry
+
+The failure handling function should return `task_need_retry` when the failed SAI call may be resolved in a subsequent attempt.
+
+4. exit(EXIT_FAILURE) -- Crash and trigger SwSS auto-restart
+
+Some of the failures can be resolved by restarting SwSS.
+In the scenario where such failures happen, the failure handling function in the orchagent should exit with `EXIT_FAILURE` and trigger SwSS auto restart.
+
+### 2.4 DB changes
+
+An ERROR_DB will be introduced to escalate the failures from orchagent to upper layers such as fpmsyncd.
+
+The schema of ERROR_DB is designed as follows:
+
+```
+ERROR_{{TABLE_TYPE}}_TABLE|entry
+    "failed_orch": {{failed orch type}}
+    "failed_SAI": {{failed SAI type}}
+    "opcode": {{failed SAI operation type}}
+    "status": {{sai_status}}
+    "attributes": "attr_type0,attr_type1,..."
+    "attr_values": "attr_value0,attr_value1,..."
+    "counter": {{count}}
+```
+
+The table and key in ERROR_DB correspond to the table and key in APPL_DB where SAI failures happen (e.g., SAI failure happens when conducting operations for APPL_DB entry `ROUTE_TABLE:0.0.0.0/0`, the corresponding key in ERROR_DB should be `ERROR_ROUTE_TABLE:0.0.0.0/0`).
+
+The field `failed_orch` indicates the type of orch where the SAI failure happens.
+
+The field `failed_SAI` indicates the type of SAI in which the SAI failure happens.
+
+The field `opcode` indicates the method that failed.
+Possible values include `CREATE/SET/DELETE`.
+
+The field `status` stores the status of the SAI operation (e.g., SAI_STATUS_NOT_SUPPORTED, SAI_STATUS_FAILURE).
+
+The ERROR_DB entry also includes a list of attributes (comma separated) and the corresponding values (comma separated) that the failed operation tries to set.
+
+The field `counter` stores the number of failures for the same entry. It could be used as a reference for handling the failure.
+
+The upstream processes are expected to consume the ERROR_DB entries and remove the handled failures from the ERROR_DB.
+Assuming the upstream processes have the proper consumption of ERROR_DB entries and failure handling logic (these are not currently available for upstreams processes and need to be added), the ERROR_DB should not keep accumulating failures in ERROR_DB and consuming memory.
```

Comment:
When an entry is deleted from APPL_DB, any ERROR_DB entries with the same key could be cleared?

Reply:
I think this should be true in most scenarios since the ERROR_DB entries are designed to track the failure in applying APPL_DB entries or basically mismatch between APPL_DB and the hardware. When an entry is removed from ERROR_DB, the mismatch between APPL_DB and the hardware should get cleared, and most likely the corresponding ERROR_DB entry is no longer needed. Yet if the upstream process believes that extra operation is needed for the failure, I think the upstream process could have the freedom to hold the entry until the operations are done.